### PR TITLE
ddl: refine the error message for the expression index for PRIMARY KEY

### LINKED ISSUE #20320: Refine the error message for the expression index for PRIMARY KEY
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
CREATE TABLE t1 (col1 INT, PRIMARY KEY ((ABS(col1))));
Error 3106: 'Defining a virtual generated column as primary key' is not supported for generated columns.
```
### 2. What did you expect to see? (Required)

The primary key cannot be an expression index.

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by m4ver1k**
### What problem does this PR solve?
Refine the error message for the expression index for PRIMARY KEY
Issue Number: close #20320 

Problem Summary:

### What is changed and how it works?

What's Changed: Error message returned when trying create a primary key on generated column,

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -99,7 +99,7 @@ func checkPKOnGeneratedColumn(tblInfo *model.TableInfo, indexPartSpecifications
 		}
 		// Virtual columns cannot be used in primary key.
 		if lastCol.IsGenerated() && !lastCol.GeneratedStored {
-			return nil, ErrUnsupportedOnGeneratedColumn.GenWithStackByArgs("Defining a virtual generated column as primary key")
```
Comment (zz-jason) [inline ddl/index.go:102]: could you add some tests?

Comment (Howie59) [reply ddl/index.go:102]: i think it is not simply change one line

Comment (wjhuang2016) [reply ddl/index.go:102]: I think the manual test is enough.

**Comment by bb7133** [comment]
LGTM

File: ddl/index.go [hunk 2]
```diff
@@ -99,7 +99,7 @@ func checkPKOnGeneratedColumn(tblInfo *model.TableInfo, indexPartSpecifications
...
+			return nil, ErrUnsupportedOnGeneratedColumn.GenWithStackByArgs("Primary Key Constraint")
```
Comment (wjhuang2016) [inline ddl/index.go:102]: The error message should be "The primary key cannot be an expression index." and it's error code should be 3756.

Comment (m4ver1k) [reply ddl/index.go:102]: Will update, thank you

Comment (wjhuang2016) [reply ddl/index.go:102]: image
The err ErrUnsupportedOnGeneratedColumn is wrong here, you should update it to ErrFunctionalIndexPrimaryKey

Comment (wjhuang2016) [reply ddl/index.go:102]: There are two possible situations. First, creating a primary key on a generated column. Two, creating a primary key on an expression index. You should take care of both of them.

Comment (wjhuang2016) [reply ddl/index.go:102]: Hint: if creating a primary key on an expression index, the col.Hidden will be true. You can use this to decide which error to return.

**Comment by wjhuang2016** [comment]
@m4ver1k Hi, could you update this PR?

**Comment by wjhuang2016** [comment]
@m4ver1k Hi, ping again, could you update this PR?

**Comment by m4ver1k** [comment]
@wjhuang2016 Sorry have been busy with work not finding time. Apologies.

**Comment by wjhuang2016** [comment]
> @wjhuang2016 Sorry have been busy with work not finding time. Apologies.

Never mind, I had fixed this issue with another PR. So you can try another issue if you are inserested.

**Comment by xhebox** [comment]
The linked issue has been fixed by #21419.

**Pull request closed without merging**